File: JOI2007_C_darts.scala
object Main {
  def main(args: Array[String]) = {
    val lines = scala.io.Source.stdin.getLines.toStream
    val Array(numLines, limit) = lines.head.split(" ").map(_.toInt)
    val scores = 0 +: lines.drop(1).map(_.toInt).toArray
    val doubles = Array.fill((numLines+1)*(numLines+1))(0)

    var max = 0
    for {
      throw1 <- scores
      throw2 <- scores
    } {
      doubles(i) = throw1 + throw2
      i += 1
    }
    java.util.Arrays.sort(doubles)

    for (double <- doubles) {
      var cand = double
      if (cand <= limit) {
        if (cand < limit) cand = cand + bin(0, doubles.length, limit - cand, doubles)
        if (cand > max) max = cand
      }
    }

    println(max)
  }

  def bin(from: Int, to: Int, limit: Int, array: Array[Int]): Int = {
    val half = (from + to) / 2
    if (from == half) return array(from)
    val cand = array(half)
    if (cand <= limit) bin(half, to, limit, array) else bin(from, half, limit, array)
  }
}

File: JOI2007_C_darts.go
package main

import "fmt"
import "sort"

func binSearch(from int, to int, limit int, inputs []int) int {
	half := (from + to) / 2
	if from == half {
		return inputs[from]
	}

	if inputs[half] <= limit {
		return binSearch(half, to, limit, inputs)
	} else {
		return binSearch(from, half, limit, inputs)
	}
}

func main() {
	var length int
	var limit int
	var max int = 0
	fmt.Scan(&length)
	length = length + 1
	fmt.Scan(&limit)
	inputs := make([]int, length)
	doubles := []int{}

	inputs[0] = 0
	for i := 1; i < length; i++ {
		fmt.Scan(&inputs[i])
	}
	for _, num1 := range inputs {
		for _, num2 := range inputs {
			doubles = append(doubles, num1+num2)
		}
	}
	sort.Ints(doubles)

	for _, value := range doubles {
		if value <= limit {
			if value < limit {
				value = value + binSearch(0, length*length, limit-value, doubles)
			}
			if max < value {
				max = value
			}
		}
	}
	fmt.Println(max)
}
